src/backend/distributed/planner/fast_path_router_planner.c

```diff
@@ -0,0 +1,391 @@
+	result->planTree = (Plan *) plan;
+
+	relationId = ExtractFirstDistributedTableId(parse);
+	result->relationOids = list_make1_oid(relationId);
+
+	return result;
+}
+
+
+/*
+ * FastPathRouterQuery gets a query and returns true if the query is eligable for
+ * being a fast path router query.
+ * The requirements for the fast path query can be listed below:
+ *
+ *   - SELECT query without CTES, sublinks-subqueries, set operations
+ *   - The query should touch only a single hash distributed or reference table
+ *   - The distribution with equality operator should be in the WHERE clause
+ *      and it should be ANDed with any other filters. Also, the distribution
+ *      key should only exists once in the WHERE clause. So basically,
+ *          SELECT ... FROM dist_table WHERE dist_key = X
+ *   - No returning for UPDATE/DELETE queries
+ */
+bool
+FastPathRouterQuery(Query *query)
+{
+	RangeTblEntry *rangeTableEntry = NULL;
+	List *rangeTableList = NIL;
+	FromExpr *joinTree = query->jointree;
+	Node *quals = NULL;
+	Oid distributedTableId = InvalidOid;
+	Var *distributionKey = NULL;
+	DistTableCacheEntry *cacheEntry = NULL;
+	List *varClauseList = NIL;
+	ListCell *varClauseCell = NULL;
+	int partitionColumnReferenceCount = 0;
+
+	if (!EnableFastPathRouterPlanner)
+	{
+		return false;
+	}
+
+	if (!(query->commandType == CMD_SELECT || query->commandType == CMD_UPDATE ||
+		  query->commandType == CMD_DELETE))
+	{
+		return false;
+	}
+
+	/*
+	 * We want to deal with only very simple select queries. Some of the
+	 * checks might be too restrictive, still we prefer this way.
+	 */
+	if (query->cteList != NIL || query->returningList != NIL ||
+		query->hasSubLinks || query->setOperations != NULL ||
+		query->hasForUpdate || query->hasTargetSRFs ||
+		query->hasModifyingCTE)
+	{
+		return false;
+	}
+
+	/* make sure that the only range table in FROM clause */
+	if (list_length(query->rtable) != 1)
+	{
+		return false;
+	}
+
+	rangeTableEntry = (RangeTblEntry *) linitial(query->rtable);
+	if (rangeTableEntry->rtekind != RTE_RELATION)
+	{
+		return false;
+	}
+
+	/*
+	 * We don't want to deal with potentially overlapping
+	 * append/range distributed tables.
+	 */
+	distributedTableId = rangeTableEntry->relid;
+	cacheEntry = DistributedTableCacheEntry(distributedTableId);
+	if (!(cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH ||
+		  cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE))
+	{
+		return false;
+	}
+
+	/* WHERE clause should not be empty for distributed tables */
+	if (joinTree == NULL ||
+		(cacheEntry->partitionMethod != DISTRIBUTE_BY_NONE && joinTree->quals == NULL))
+	{
+		return false;
+	}
+
+	/*
+	 * Pull all range table entries. We prefer this to make sure that
+	 * there is no subqueries in the any part of the query, including
+	 * the from clause, where clause, target list or HAVING clause etc.
+	 */
+	ExtractRangeTableEntryWalker((Node *) query, &rangeTableList);
+	if (list_length(rangeTableList) != 1)
```

Comment:
This seems a bit inefficient for what it's trying to do. In the spirit of this PR, I wonder whether there's a faster implementation.

Reply:
Also, hasSublinks should catch subqueries in the where/target list/having, shouldn't it?